docs/design/offline-cascading-deletes/README.md

```diff
@@ -0,0 +1,414 @@
+            natural key changes, with the language instruction `allow primary
+            key updates`. Thus it will be possible, in the API, to quickly
+            detect which resources are allowed to use the process described
+            above.
+
+            1. Yes - proceed to "START CASCADE UNIT OF WORK" below
+            2. No - update the document _except for the natural key_. Just
+               ignore the natural key update.
+    2. START CASCADE **UNIT OF WORK**:
+        1. Start atomic transaction
+        2. Update the document as requested, including the new Meadowlark ID.
+        3. Update all incoming references to hold the new Meadowlark ID, with
+           code something like:
+
+            ```none
+            # MongoDB
+            db.getCollection('documents').updateMany(
+              { 'outboundRefs': '<OLD VALUE>' },
+              { $set: { 'outboundRefs.$[elem]': '<NEW VALUE>' } },
+              {
+                upsert: false,
+                writeConcern: <CONFIG VALUE>,
+                arrayFilters: [ { 'elem': { $eq: '<OLD VALUE>' } } ]
+              }
+            )
+
+            # PostgreSQL
+            update references set parent_document_id = '<NEW VALUE>'
+              where parent_document_id = '<OLD VALUE>';
+            ```
+
+        4. Write a record to the new Outbox collection/table, like:
+
+              ```json
+              {
+                "eventType": "key-update",
+                "event": {
+                  "oldMeadowlarkId": <OLD VALUE>,
+                  "newMeadowlarkId": <NEW VALUE>,
+                  "newIdentity": {
+                    "sKeyPart1": "a",
+                    "sKeyPart2": "new-value",
+                  }
+                },
+                "timestamp": <current Unix timestamp>,
+                "id": <UUID v1>,
+                "partitionKey": <UUID of changed document>
+              }
+              ```
+
+        5. Commit transaction
+
+### Dispatch
+
+Events will be dispatched from the outbox to a message broker / queue through
+changed data capture (CDC) using Debezium. See [Meadowlark Streaming and
+Downstream Data Stores](https://techdocs.ed-fi.org/x/UgWgBw) for more
+information.
+
+### Offline Process - Event Handler / Stream Processor
+
+There will be a monitoring service / function (details forthcoming) that detects
+the `key-update` event and then takes action to fix dependencies. Naive approach:
+
+1. For each document that references the `event.documentUuid` ("child document"):
+   1. For each component of `event.newIdentity`: replace child document's value
+      for matching key.
+   2. Calculate the child document's new Meadowlark ID and update the document.
+   3. Create a `key-update` event _for the child document_, so that any
+      "grandchildren" of the original change will also receive the update.
+   4. Create a `document-update` event for the child document, for updating
+      OpenSearch (details forthcoming).
```

Comment:
Probably don't need (4), because OpenSearch would be updated from the events triggered by the `documents` collection/table.